server/src/main/java/org/elasticsearch/transport/TransportDecompressor.java

```diff
@@ -8,130 +8,59 @@
 
-import org.apache.lucene.util.BytesRef;
-import org.apache.lucene.util.BytesRefIterator;
-import org.elasticsearch.common.bytes.BytesArray;
 import org.elasticsearch.common.bytes.BytesReference;
 import org.elasticsearch.common.bytes.ReleasableBytesReference;
-import org.elasticsearch.common.compress.CompressorFactory;
-import org.elasticsearch.core.Releasable;
-import org.elasticsearch.common.recycler.Recycler;
+import org.elasticsearch.common.compress.DeflateCompressor;
 import org.elasticsearch.common.util.PageCacheRecycler;
+import org.elasticsearch.core.Releasable;
 
 import java.io.IOException;
-import java.util.ArrayDeque;
-import java.util.zip.DataFormatException;
-import java.util.zip.Inflater;
 
-public class TransportDecompressor implements Releasable {
+public interface TransportDecompressor extends Releasable {
 
-    private final Inflater inflater;
-    private final PageCacheRecycler recycler;
-    private final ArrayDeque<Recycler.V<byte[]>> pages;
-    private int pageOffset = PageCacheRecycler.BYTE_PAGE_SIZE;
-    private boolean hasReadHeader = false;
+    int decompress(BytesReference bytesReference) throws IOException;
 
-    public TransportDecompressor(PageCacheRecycler recycler) {
-        this.recycler = recycler;
-        inflater = new Inflater(true);
-        pages = new ArrayDeque<>(4);
-    }
+    ReleasableBytesReference pollDecompressedPage(boolean isEOS);
 
-    public int decompress(BytesReference bytesReference) throws IOException {
-        int bytesConsumed = 0;
-        if (hasReadHeader == false) {
-            if (CompressorFactory.COMPRESSOR.isCompressed(bytesReference) == false) {
-                int maxToRead = Math.min(bytesReference.length(), 10);
-                StringBuilder sb = new StringBuilder("stream marked as compressed, but no compressor found, first [")
-                    .append(maxToRead).append("] content bytes out of [").append(bytesReference.length())
-                    .append("] readable bytes with message size [").append(bytesReference.length()).append("] ").append("] are [");
-                for (int i = 0; i < maxToRead; i++) {
-                    sb.append(bytesReference.get(i)).append(",");
-                }
-                sb.append("]");
-                throw new IllegalStateException(sb.toString());
-            }
-            hasReadHeader = true;
-            int headerLength = CompressorFactory.COMPRESSOR.headerLength();
-            bytesReference = bytesReference.slice(headerLength, bytesReference.length() - headerLength);
-            bytesConsumed += headerLength;
+    @Override
+    void close();
+
+    static TransportDecompressor getDecompressor(PageCacheRecycler recycler, BytesReference bytes) throws IOException {
+        if (bytes.length() < DeflateCompressor.HEADER.length) {
+            return null;
+        }
+        byte firstByte = bytes.get(0);
+        byte[] header;
+        if (firstByte == CompressionScheme.DEFLATE_HEADER[0]) {
```

Comment:
I wonder if we should just interpret the header as `int`, then we don't have to have a mutable `public static byte[]` and can just `switch` on that int here (faster and less code)?